presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java

```diff
@@ -0,0 +1,281 @@
+
+        @Override
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
+        {
+            node = (FilterNode) super.visitFilter(node, context);
+
+            List<InPredicate> inPredicates = collectInPredicates(ImmutableList.of(node.getPredicate()));
+            ImmutableMap.Builder<InPredicate, Expression> inPredicateMapping = ImmutableMap.builder();
+            node = rewriteInPredicates(node, inPredicates, inPredicateMapping);
+            Expression expression = replaceExpression(node.getPredicate(), inPredicateMapping.build());
+            return new FilterNode(node.getId(), getOnlyElement(node.getSources()), expression);
+        }
+
+        @Override
+        public PlanNode visitProject(ProjectNode node, RewriteContext<Void> context)
+        {
+            node = (ProjectNode) super.visitProject(node, context);
+
+            List<InPredicate> inPredicates = collectInPredicates(node.getAssignments().values());
+            ImmutableMap.Builder<InPredicate, Expression> inPredicateMapping = ImmutableMap.builder();
+            node = rewriteInPredicates(node, inPredicates, inPredicateMapping);
+            return new ProjectNode(node.getId(),
+                    getOnlyElement(node.getSources()),
+                    rewriteAssignments(node, inPredicateMapping.build()));
+        }
+
+        private <T extends PlanNode> T rewriteInPredicates(PlanNode node, List<InPredicate> inPredicates, ImmutableMap.Builder<InPredicate, Expression> inPredicateMapping)
+        {
+            for (InPredicate inPredicate : inPredicates) {
+                InsertSemiJoinContext insertSemiJoinContext = new InsertSemiJoinContext(inPredicate);
+                node = rewriteWith(new InsertSemiJoinRewriter(idAllocator, symbolAllocator), node, insertSemiJoinContext);
+                inPredicateMapping.putAll(insertSemiJoinContext.getInPredicateMapping());
+            }
+            return (T) node;
+        }
+
+        private ImmutableMap<Symbol, Expression> rewriteAssignments(ProjectNode node, ImmutableMap<InPredicate, Expression> inPredicateMapping)
+        {
+            ImmutableMap.Builder<Symbol, Expression> assignmentsBuilder = ImmutableMap.builder();
+            Map<Symbol, Expression> assignments = node.getAssignments();
+            for (Symbol symbol : assignments.keySet()) {
+                assignmentsBuilder.put(symbol, replaceExpression(assignments.get(symbol), inPredicateMapping));
+            }
+            return assignmentsBuilder.build();
+        }
+
+        private List<InPredicate> collectInPredicates(Collection<Expression> expressions)
+        {
+            ImmutableList.Builder<InPredicate> listBuilder = ImmutableList.builder();
+            for (Expression expression : expressions) {
+                new DefaultTraversalVisitor<Void, Void>()
+                {
+                    @Override
+                    protected Void visitInPredicate(InPredicate node, Void context)
```

Comment:
This will also include any predicates of the form `xxx in (a, b, c, ...)`. Is that intentional? Isn't this supposed to do it only for cases where the "value expression" of the IN predicate is a subquery?